lib/workers/pr/changelog/source-gitlab.ts

```diff
@@ -0,0 +1,183 @@
+import URL from 'url';
+import { api } from '../../../platform/github/gh-got-wrapper';
+import { logger } from '../../../logger';
+import * as hostRules from '../../../util/host-rules';
+import * as allVersioning from '../../../versioning';
+import { addReleaseNotes } from './release-notes';
+import { ChangeLogError, ChangeLogRelease, ChangeLogResult } from './common';
+import { Release } from '../../../datasource';
+import { PLATFORM_TYPE_GITLAB } from '../../../constants/platforms';
+import { BranchUpgradeConfig } from '../../common';
+
+const { get: ghGot } = api;
+
+async function getTags(
+  endpoint: string,
+  versionScheme: string,
+  repository: string
+): Promise<string[]> {
+  logger.trace('getTags() from gitlab');
+  let url = endpoint
+    ? endpoint.replace(/\/?$/, '/')
+    : /* istanbul ignore next: not possible to test, maybe never possible? */ 'https://gitlab.com/api/v4/';
+  const repoid = repository.replace(/\.git/, '').replace(/\//g, '%2F');
+  url += `projects/${repoid}/repository/tags`;
+  try {
+    const res = await ghGot<{ name: string }[]>(url, {
+      paginate: true,
+    });
+
+    const tags = (res && res.body) || [];
+
+    if (!tags.length) {
+      logger.debug({ repoid }, 'repository has no Gitlab tags');
+    }
+
+    return tags.map(tag => tag.name).filter(Boolean);
+  } catch (err) {
+    logger.info({ sourceRepo: repoid }, 'Failed to fetch Gitlab tags');
+    // istanbul ignore if
+    if (err.message && err.message.includes('Bad credentials')) {
+      logger.warn('Bad credentials triggering tag fail lookup in changelog');
+      throw err;
+    }
+    return [];
+  }
+}
+
+export async function getChangeLogJSON({
+  endpoint,
+  versioning,
+  fromVersion,
+  toVersion,
+  sourceUrl,
+  releases,
+  depName,
+  manager,
+}: BranchUpgradeConfig): Promise<ChangeLogResult | null> {
+  if (sourceUrl === 'https://gitlab.com/DefinitelyTyped/DefinitelyTyped') {
+    logger.debug('No release notes for @types');
+    return null;
+  }
+  logger.trace('getChangeLogJSON for gitlab');
+  const version = allVersioning.get(versioning);
+  const { protocol, host, pathname } = URL.parse(sourceUrl);
+  logger.debug({ protocol, host, pathname }, 'Protocol, host, pathname');
+  const baseURL = 'https://gitlab.com/';
+  const url = sourceUrl;
+  const config = hostRules.find({
+    hostType: PLATFORM_TYPE_GITLAB,
+    url,
+  });
+  // istanbul ignore if
+  if (!config.token) {
+    // prettier-ignore
+    if (URL.parse(sourceUrl).host.search(/gitlab/) !== -1) { // lgtm [js/incomplete-url-substring-sanitization]
+      logger.warn(
+        {manager, depName, sourceUrl},
+        'No gitlab token has been configured. Skipping release notes retrieval'
+      );
+      return {error: ChangeLogError.MissingGitlabToken};
```

Comment:
This behaviour mimics what's being done for GitHub, but we are not sure if it should be done for gitlab.com. Could use some feedback here.